extensions/gitpod-web/src/extension.ts

```diff
@@ -400,85 +511,184 @@ export class GitpodWorkspaceTreeDataProvider implements vscode.TreeDataProvider<
+		const scriptUri = webview.asWebviewUri(vscode.Uri.joinPath(this.context.extensionUri, 'resources', 'ports/bundle.js'));
+		const styleUri = webview.asWebviewUri(vscode.Uri.joinPath(this.context.extensionUri, 'resources', 'ports/bundle.css'));
+		const nonce = getNonce();
+		return `<!DOCTYPE html>
+            <html lang="en">
+            <head>
+                <meta charset="UTF-8" />
+                <meta http-equiv="X-UA-Compatible" content="IE=edge" />
+                <meta
+		        http-equiv="Content-Security-Policy"
+		        content="default-src 'none'; img-src data: ${webview.cspSource}; style-src ${webview.cspSource} 'nonce-${nonce}'; script-src 'nonce-${nonce}';"
+		        />
+                <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+
+                <link href="${styleUri}" rel="stylesheet" />
+                <script nonce="${nonce}">
+                    const vscode = acquireVsCodeApi();
+                </script>
+                <title>Gitpod Port View</title>
+            </head>
+            <body></body>
+            <script nonce="${nonce}" src="${scriptUri}"></script>
+            </html>`;
+	}
+
+	readonly tunnels = new Map<number, vscode.TunnelDescription>();
+	updateTunnels(tunnels: vscode.TunnelDescription[]): void {
+		this.tunnels.clear();
+		for (const tunnel of tunnels) {
+			this.tunnels.set(tunnel.remoteAddress.port, tunnel);
+		}
+		this.update();
+	}
+
+	private portStatus: PortsStatusResponse | undefined;
+	updatePortsStatus(portsStatus: PortsStatusResponse): void {
+		this.portStatus = portsStatus;
+		this.update();
+	}
+
+	private updating = false;
+	private update(): void {
+		if (this.updating) { return; }
+		this.updating = true;
+		try {
+			if (!this.portStatus) { return; }
+			this.portStatus.getPortsList().forEach(e => {
+				const localPort = e.getLocalPort();
+				const gitpodPort = this.portMap.get(localPort);
+				const tunnel = this.tunnels.get(localPort);
+				if (!gitpodPort) {
+					this.portMap.set(localPort, new GitpodWorkspacePort(localPort, this.context, e, tunnel));
+					return;
+				}
+				const prevStatus = gitpodPort.status;
+				gitpodPort.update(e, tunnel);
+				if (isExposedServedGitpodWorkspacePort(gitpodPort) && !isExposedServedPort(prevStatus)) {
+					this.onDidExposeServedPortEmitter.fire(gitpodPort);
+				}
+			});
+			this.onDidChangePortsEmitter.fire(this.portMap);
+			this.updateHtml();
+		} finally {
+			this.updating = false;
+		}
+	}
+
+	private updateHtml(): void {
+		const ports = Array.from(this.portMap.values()).map(e => e.toSvelteObject());
+		this._view?.webview.postMessage({ command: 'updatePorts', ports });
+	}
+
+	private onHtmlCommand() {
+		this._view?.webview.onDidReceiveMessage(async (message: { command: PortCommand; port: { info: PortInfo; status: PortsStatus.AsObject } }) => {
+			if (message.command === 'queryPortData') {
+				this.updateHtml();
+				return;
+			}
```

Comment:
We need this because `onDidChangeVisibility` does not work sometimes... 

That's why [Anton](https://github.com/gitpod-io/openvscode-server/pull/361#issuecomment-1152110710) needs to trigger a ports update to load this Webview first time